api/src/main/java/org/apache/iceberg/actions/CompactDataFiles.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.Map;
+import org.apache.iceberg.actions.compaction.BinPack;
+import org.apache.iceberg.expressions.Expression;
+
+public interface CompactDataFiles extends Action<CompactDataFiles, CompactDataFiles.Result> {
```

Comment:
Do we want to offer this under a new name? I thought this would swap the `RewriteDataFiles` implementation in the new actions API. Right now, `RewriteDataFiles` from the new API is not implemented so we don't have to update any other place, can just move the new logic there.

Plus, I tend to think `rewrite` fits slightly better as we not necessarily compact. We can split files or recluster them, for example.

Reply:
I think we should go for Optimize, I feel like Rewrite is a bit difficult because a successful run of this command can end up not changing anything. Which I think makes more sense for Optimize which if it does nothing, implies it is already optimized.

Reply:
I am flexible here. I think `RewriteDataFiles` was fine as it was generic enough to indicate we are rewriting data files without changing data in them. In addition, users are already familiar with the existing name. Not a strong opinion. If we are to change the name, now is the time to do this.

As this action can be used to rewrite data with another output spec, using `OptimizeDataFiles` may not fit very well.

Reply:
That reminds me that we need an option for the output spec.

Reply:
I was wondering if we really need the output spec. Shouldn't we always be targeting the latest spec? and if someone wants to change spec and then optimize, then they should do it as two operations?

Reply:
I don't think it will be common but it may be a valid use case to have multiple specs in a table. For example, the volume of data for old partitions may be much smaller so we may actually want to keep the old spec for some parts of the data.

Reply:
I also thought a bit more about the naming. I think I'd still go with the old `RewriteDataFiles` name.

Definitely not a strong opinion, though.

Reply:
Ok if no one else has comments i'll have this be RewriteDataFiles and we will rename the old one to v1